### ISSUE TITLE
linux-headers-5.4.0-90-generic missing

### ISSUE THREAD

**Issue by stauder** [MrSimonEmms/gitpod-k3s-guide#1]
I have this problem:

> kubectl logs ws-daemon-2nd7s -c shiftfs-module-loader
+ mkdir -p /lib/modules/5.4.0-90-generic
+ ln -s /usr/src_node/linux-headers-5.4.0-90-generic /lib/modules/5.4.0-90-generic/build
+ dkms install -m shiftfs -v 1.2 -k 5.4.0-90-generic --kernelsourcedir /usr/src_node/linux-headers-5.4.0-90-generic

Creating symlink /var/lib/dkms/shiftfs/1.2/source ->
                 /usr/src/shiftfs-1.2

DKMS: add completed.
Error! Your kernel headers for kernel 5.4.0-90-generic cannot be found.
Please install the linux-headers-5.4.0-90-generic package,
or use the --kernelsourcedir option to tell DKMS where it's located

**Comment by mrsimonemms** [comment]
@stauder can you add the following to your `setup.sh` script and see if this solves it for you?

```shell
echo "Install linux-headers"
ssh "${USER}@${IP}" "sudo apt-get update"
ssh "${USER}@${IP}" "sudo apt-get install -y linux-headers-$(uname -r) linux-headers-generic"
```

I've created PR #2 for this, so you can just pull that branch for ease. I think this will solve your issue, although I've found another issue on a GCP machine.

Out of interest, where is your machine?

**Comment by stauder** [comment]
Hello @MrSimonEmms 

For me it also works with this fix now.

I try this on a virtual server at Hetzer with Ubuntu 20.04.3 LTS minimal .

**In Gitpod:**
1) I opened a Gitpod workspace out of https://github.com/MrSimonEmms/gitpod-k3s-guide
2) Generated SSH keys with ssh-keygen
3) cp .env.example to .env
4) Changed .env corresponding to the setup, especially USER="gitpod" . I used release-2022.01.10

**On the virtual server**
1) adduser gitpod
2) Add to /etc/sudoers
gitpod ALL=(ALL) NOPASSWD:ALL
3) Add the key from Gitpod Workspace ~gitpod/.ssh/id_rsa.pub to ~gitpod/.ssh/authorized_keys

Now the workspace on my server starts! Thanks a lot for your fix!

During installation I see some warnings, that can most probably be ignored:

[...]
Get:5 http://de.archive.ubuntu.com/ubuntu focal-updates/main amd64 linux-headers-generic amd64 5.4.0.99.103 [2,436 B]
debconf: unable to initialize frontend: Dialog
debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.30.0 /usr/local/share/perl/5.30.0 /usr/lib/x86_64-linux-gnu/perl5/5.30 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.30 /usr/share/perl/5.30 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 7, <> line 5.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Fetched 27.0 MB in 2s (17.5 MB/s)
Selecting previously unselected package linux-gcp-5.13-headers-5.13.0-1013.
[...]

**Comment by mrsimonemms** [comment]
Thanks for checking @stauder.

There appears to be another issue that I've come across with VMs created in GCP/Azure. It's in the same `shiftfs-module-loader` init container

Terminal:

```
Creating symlink /var/lib/dkms/shiftfs/1.2/source ->
                 /usr/src/shiftfs-1.2

DKMS: add completed.

Preparing kernel 5.11.0-1029-gcp for module build:
(This is not compiling a kernel, just preparing kernel symbols)
Storing current .config to be restored when complete
Running Generic preparation routine
make mrproper....(bad exit status: 2)
/usr/sbin/dkms: line 1211: .config: Read-only file system
Warning: using /usr/src_node/linux-headers-5.11.0-1029-gcp/.config
(I hope this is the correct config for this kernel)
make oldconfig...(bad exit status: 2)
make prepare-all...(bad exit status: 2)

Building module:
cleaning build area...
make -j2 KERNELRELEASE=5.11.0-1029-gcp -C /usr/src_node/linux-headers-5.11.0-1029-gcp M=/var/lib/dkms/shiftfs/1.2/build...(bad exit status: 2)
Error! Bad return status for module build on kernel: 5.11.0-1029-gcp (x86_64)
Consult /var/lib/dkms/shiftfs/1.2/build/make.log for more information.
```

Logs:
```
cat /var/lib/dkms/shiftfs/1.2/build/make.log 
DKMS make.log for shiftfs-1.2 for kernel 5.11.0-1029-gcp (x86_64)
Sun Feb 13 13:28:14 UTC 2022
make: Entering directory '/usr/src_node/linux-headers-5.11.0-1029-gcp'
  CC [M]  /var/lib/dkms/shiftfs/1.2/build/shiftfs.o
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c: In function 'shiftfs_fiemap':
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c:727:13: error: dereferencing pointer to incomplete type 'struct fiemap_extent_info'
  if (fieinfo->fi_flags & FIEMAP_FLAG_SYNC)
             ^~
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c:727:26: error: 'FIEMAP_FLAG_SYNC' undeclared (first use in this function); did you mean 'FS_XFLAG_SYNC'?
  if (fieinfo->fi_flags & FIEMAP_FLAG_SYNC)
                          ^~~~~~~~~~~~~~~~
                          FS_XFLAG_SYNC
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c:727:26: note: each undeclared identifier is reported only once for each function it appears in
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c: In function 'shiftfs_btrfs_ioctl_fd_restore':
/var/lib/dkms/shiftfs/1.2/build/shiftfs.c:1422:2: error: implicit declaration of function '__close_fd'; did you mean 'close_fd'? [-Werror=implicit-function-declaration]
  __close_fd(current->files, fd);
  ^~~~~~~~~~
  close_fd
cc1: some warnings being treated as errors
scripts/Makefile.build:288: recipe for target '/var/lib/dkms/shiftfs/1.2/build/shiftfs.o' failed
make[1]: *** [/var/lib/dkms/shiftfs/1.2/build/shiftfs.o] Error 1
Makefile:1849: recipe for target '/var/lib/dkms/shiftfs/1.2/build' failed
make: *** [/var/lib/dkms/shiftfs/1.2/build] Error 2
make: Leaving directory '/usr/src_node/linux-headers-5.11.0-1029-gcp'
```

**Comment by mrsimonemms** [comment]
@stauder In fact, I'm going to raise this as a separate issue. I've just merged #3 which changes the `.env` file (change `IP` to `IP_LIST`)

**Comment by stauder** [comment]
Hello @MrSimonEmms 
In my eyes, this fix has the problem, that (uname -r) gives back the OS of the installing host, but not the remote one. For me this results in
Unable to locate package linux-headers-5.13.0-1015-gcp

I'm having 5.4.0-100-generic , is this wrong?

**Comment by mrsimonemms** [comment]
My understanding is that they're the same thing - because of the way it's mapped, the kernel version is the same on the host and container image.

I'm a little confused as to what's going on here. However, since yesterday's [2022.02.0 release](https://github.com/gitpod-io/gitpod/releases/tag/2022.02.0), we've fixed the issues with `fuse` - can you try swapping `shiftfs` for `fuse`?

**Comment by stauder** [comment]
I run the installer in a gitpod workspace. There uname returns 5.13.0-1015-gcp . But on my remote server 5.4.0-100-generic is installed and there is no package for 5.13.0-1015-gcp .

What fixed it was
```
    UNAME=`ssh "${USER}@${IP}" uname -r`
    ssh "${USER}@${IP}" "sudo apt-get install -y linux-headers-${UNAME} linux-headers-generic"

```
But I guess you mean the linux-headers are not needed at all anymore with fuse?

**Comment by mrsimonemms** [comment]
I'll do some testing on that - I tried it on both Azure and GCP and my fix works, but clearly didn't for you which is strange.

The linux-headers are purely for `shiftfs` (I fairly confident it's [this section](https://github.com/gitpod-io/gitpod/blob/main/install/installer/pkg/components/ws-daemon/daemonset.go#L97)). Switching to `fuse` should negate the need for them

**Comment by mrsimonemms** [comment]
Fixed by #8
